core/ppu_bg.go

```diff
@@ -176,11 +188,107 @@ func (ppu *PPU) tileFromBackground(background uint8, x uint16, y uint16) tile {
 	// See: https://wiki.superfamicom.org/backgrounds
 	raw := bit.JoinUint16(ppu.vram.bytes[addr], ppu.vram.bytes[addr+1])
 
-	return tile{
+	hSize, vSize := bg.tileSize()
+	colorDepth := ppu.colorDepth(background)
+	tileNumber := raw & 0x3FF
+
+	return bgTile{
+		baseTile: baseTile{
+			palette:    uint8((raw >> 10) & 0x7),
+			addr:       uint16(bg.tileSetBaseAddr)<<13 + uint16(tileNumber)*baseTileSize(colorDepth),
+			colorDepth: ppu.colorDepth(background),
+		},
 		vFlip:    raw&0x8000 != 0,
 		hFlip:    raw&0x4000 != 0,
 		priority: raw&0x2000 != 0,
-		palette:  uint8((raw >> 10) & 0x7),
-		number:   raw & 0x3ff,
+		hSize:    hSize,
+		vSize:    vSize,
+	}
+}
+
+//tileSize returns the size in pixel of tiles in the background
+func (bg *bg) tileSize() (uint16, uint16) {
+	hSize, vSize := uint16(8), uint16(8)
+	if bg.tileSizeFlag {
+		hSize, vSize = 16, 16
+	}
+
+	return hSize, vSize
+}
+
+// colorDepth returns the number of bits used for the colors in the background
+func (ppu *PPU) colorDepth(background uint8) uint8 {
+	panicString := "in mode %d, only background 1,2,3 are valid, attempted to use background %d"
+	switch ppu.backgroundData.screenMode {
+	case 0:
+		return 2
+	case 1:
+		switch background {
+		case 0, 1:
+			return 4
+		case 2:
+			return 2
+		default:
+			panic(fmt.Sprintf(panicString, ppu.backgroundData.screenMode, background+1))
+		}
+	case 2:
+		switch background {
+		case 0, 1:
+			return 4
+		default:
+			panic(fmt.Sprintf(panicString, ppu.backgroundData.screenMode, background+1))
+		}
+	case 3, 4:
+		switch background {
+		case 0:
+			return 8
+		case 1:
+			return 4
+		default:
+			panic(fmt.Sprintf(panicString, ppu.backgroundData.screenMode, background+1))
+		}
+	case 5:
+		switch background {
+		case 0:
+			return 4
+		case 1:
+			return 2
+		default:
+			panic(fmt.Sprintf(panicString, ppu.backgroundData.screenMode, background+1))
+		}
+	case 6:
+		switch background {
+		case 0:
+			return 4
+		default:
+			panic(fmt.Sprintf(panicString, ppu.backgroundData.screenMode, background+1))
+		}
+	case 7:
+		switch background {
+		case 0:
+			return 8
+		}
+	default:
```

Comment:
nit: should we move this outside the switch (to really ensure that the `return 0` never happens)